[AHM] Claims and Crowdloan tests

eugypalu opened the pull request:
Add pre/post migration checks to claims and crowdloan

- [x] Does not require a CHANGELOG entry

muharem commented on pallets/rc-migrator/src/crowdloan.rs:
```diff
@@ -299,3 +299,47 @@ pub fn num_leases_to_ending_block<T: Config>(num_leases: u32) -> Result<BlockNum
+				})
+				.collect();
+
+		assert!(current_map.is_empty(), "Current crowdloan data should be empty after migration");
```
should not `pallet_crowdloan::Funds` also be empty?

muharem commented on pallets/ah-migrator/src/crowdloan.rs:
```diff
@@ -165,3 +166,42 @@ where
+		assert_eq!(
+			rc_pre, all_post,
+			"Crowdloan data mismatch: Asset Hub data differs from original Relay Chain data"
+		);
```
should we also check `RcLeaseReserve` and `RcCrowdloanReserve` reserves?

eugypalu replied:
You're right, I fixed both comments, thanks!

muharem commented:
@ggwpez we need your eyes on this, when looks good to you, feel free to merge

ggwpez commented on pallets/ah-migrator/src/crowdloan.rs:
```diff
@@ -165,3 +166,83 @@ where
+impl<T: Config> crate::types::AhMigrationCheck for CrowdloanMigrator<T> {
+	type RcPrePayload = Vec<(
+		ParaId,
+		Vec<(BlockNumberFor<T>, <T as frame_system::Config>::AccountId, BalanceOf<T>)>,
```
Docs please what the elements of the tuple mean.

ggwpez commented on pallets/ah-migrator/src/crowdloan.rs:
```diff
@@ -165,3 +166,83 @@ where
+			.collect();
+
+		let lease_reserves_post: BTreeMap<_, _> = pallet_ah_ops::RcLeaseReserve::<T>::iter()
+			.map(|((block_number, para_id, account_id), amount)| {
```
Maybe a function would be good that compares two iterators, or one iterator and a vector or something.  
Otherwise this is lots of copy&paste code.

eugypalu replied:
Done, thanks! Added a verify_reserves helper to reduce code, wdyt?

ggwpez commented on pallets/rc-migrator/src/crowdloan.rs:
```diff
@@ -299,3 +299,50 @@ pub fn num_leases_to_ending_block<T: Config>(num_leases: u32) -> Result<BlockNum
+							.map(|(contributor, (amount, block_number_bytes))| {
+								let block_number: BlockNumberFor<T> =
+									BlockNumberFor::<T>::decode(&mut &block_number_bytes[..])
+										.unwrap_or_default();
```
ditto

ggwpez commented on pallets/ah-migrator/src/crowdloan.rs:
```diff
@@ -165,3 +169,80 @@ where
+	///   - `BlockNumberFor<T>`: The block number at which this deposit can be unreserved
+	///   - `AccountId`: The depositor account
+	///   - `BalanceOf<T>`: The reserved amount
+	type RcPrePayload = Vec<(ParaId, Vec<(BlockNumberFor<T>, AccountIdOf<T>, BalanceOf<T>)>)>;
```
Could also be a struct for easier readability.

ggwpez commented on pallets/ah-migrator/src/crowdloan.rs:
```diff
@@ -165,3 +169,80 @@ where
+		);
+		verify_reserves::<T, _>(
+			&reserves_pre,
+			pallet_ah_ops::RcCrowdloanReserve::<T>::iter(),
```
Code looks nice now, thanks.

ggwpez commented on pallets/rc-migrator/src/crowdloan.rs:
```diff
@@ -299,3 +299,73 @@ pub fn num_leases_to_ending_block<T: Config>(num_leases: u32) -> Result<BlockNum
+	///   - `BalanceOf<T>`: The reserved amount
+	type RcPrePayload = Vec<(ParaId, Vec<(BlockNumberFor<T>, AccountIdOf<T>, BalanceOf<T>)>)>;
+
+	fn pre_check() -> Self::RcPrePayload {
+		// Iterate through all crowdloan funds and collect their contributions
+		pallet_crowdloan::Funds::<T>::iter()
+			.map(|(para_id, fund)| {
+				// For each fund, collect all contributions into a vector
+				let contributions =
+					pallet_crowdloan::Pallet::<T>::contribution_iterator(fund.fund_index)
+						.map(|(contributor, (amount, encoded_block_number))| {
+							// Decode the block number from bytes, defaulting to 0 if decoding fails
+							let block_number =
+								BlockNumberFor::<T>::decode(&mut &encoded_block_number[..])
+									.unwrap_or_default();
+
+							// Create a tuple of (block_number, contributor, amount)
+							// Convert amount to the expected balance type, defaulting to 0 if
+							// conversion fails
+							(block_number, contributor, amount.try_into().unwrap_or_default())
+						})
+						.collect();
+
+				// Return tuple of parachain ID and its contributions
+				(para_id, contributions)
+			})
+			.collect()
+	}
```
We can still merge it, but in the future please dont nest the code so deeply. It is difficult to read.  
You can always set up a mutable vector and push into it with a for loop.

ggwpez commented:
PS looks like CI is red.

eugypalu commented:
> PS looks like CI is red.

FIxed, now the the CI is green. Fixed the tests and a small issue in the migration that was preventing some `CrowdloanContribution` from being properly processed
